RFCs/0000-pdk-template-v2.md

```diff
@@ -0,0 +1,296 @@
+> TODO: Does files_init EVEN make sense?!??!
+> Actually it does it in an `always_apply` template
+
+> TODO: What about object_templates?
+
+Note - `/object_templates/*.erb` will still be consumed by the V2 Templating Engine.
+
+> TODO: What about `/rubocop/` ?
+
+### `<name>/template.yml`
+
+| YAML Element | Description |
+| ------------ | ----------- |
+| name (required) | A short descriptive name of the template |
+| description | A long (even multiline) description of the template. Default is the name of the template |
+| type | What type of item this is a template for. Currently only `module` is supported, but in time `control_repo` and others may be added. Default is `module` |
+| evaluate_ruby_files | An array of files to [evaluate](https://ruby-doc.org/core-2.6.4/Kernel.html#method-i-eval) prior to rendering ERB templates. This is used to inject complex logic into the templating system without the troubles of having complex logic in ERB.  For example complex branching based on settings. Default is empty array |
+| default_settings | A YAML hash of the default settings for the template. Default is empty hash |
+| always_apply | Always apply this template. If not specified in the template list, it will be applied last. Default is `false` |
+| tags | An array of string tags for this template. Always includes the directory name of the module. Default is an empty array |
+
+### `<name>/template_schema.json`
+
+### Example templates
+
+#### A template for a default module file
+
+The `.pdkignore` file will be used in all PDK compatible modules therefore it should always be applied.
+
+File layout
+
+``` text
+pdkignore/
+  +- template.yml
+  +- template_schema.json
+  +- files/
+       +- .pdkignore.erb
+```
+
+
+`template.yml`
+
+``` yaml
+---
+name: "PDK Ignore"
+description: >-
+  The PDK Ignore template manages the .pdkignore file for Puppet Modules
+type: module
+always_apply: true
+tags: ignore default
+default_settings:
+  required:
+    - '/appveyor.yml'
+    - '/.fixtures.yml'
+    - '/Gemfile'
+    - '/.gitattributes'
+    - '/.gitignore'
+    - '/.gitlab-ci.yml'
+    - '/.pdkignore'
+    - '/Rakefile'
+    - '/rakelib/'
+    - '/.rspec'
+    - '/.rubocop.yml'
+    - '/.travis.yml'
+    - '/.yardopts'
+    - '/spec/'
+    - '/.vscode/'
+    # ... many other paths
+  paths: ~
+```
+
+`template_schema.json`
+
+``` json
+{
+  "$schema": "http://json-schema.org/draft-06/schema#",
+  "$id": "http://puppet.com/schema/does_not_exist.json",
+  "type": "object",
+  "title": "The PDK Ignore Template Schema",
+  "properties": {
+    "paths": {
+      "$id": "#/properties/paths",
+      "title": "Paths property",
+      "description": "Array of file paths to also ignore by the PDK, for example, when building a module",
+      "type": "array",
+      "items": {
+        "type": "string"
+      },
+    },
+  },
+  "definitions": {}
+}
+```
+
+**Note** that the `required` setting is not documented, or has a schema, as it's internal to the template.
```

Comment:
This came originally from the fact that the modulesync configs do not provide a way to amend array- or hash-valued settings. Can we improve on that today?

Reply:
This is probably a bad example.  PDK templates already support knockout prefix which helps there.